common/go-utils/src/vrs.rs

```diff
@@ -1,76 +1,74 @@
+use std::fmt::{self};
 
 /// `Requirement` is a wrapper around `semver::Requirement` that adds
-/// - `Deserialize` and `Serialize` traits
 /// - Ability to parse go-flavored requirements
 ///
 /// The derived `Default` implementation creates a wildcard version `Requirement`.
-#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq, Eq)]
-#[serde(try_from = "String", into = "String")]
-pub struct Requirement(semver::VersionReq);
+#[derive(Debug, Default, PartialEq, Clone)]
+pub struct GoRequirement(semver::VersionReq);
 
 #[derive(thiserror::Error, Debug)]
 #[error("Couldn't parse Go version requirement: {0}")]
 pub struct RequirementParseError(#[from] semver::Error);
 
-impl Requirement {
+pub trait Version {}
+
+impl Version for GoVersion {}
+pub trait VersionRequirement<T> {
+    fn satisfies(&self, version: &T) -> bool;
+
     /// Parses a semver requirement `&str` as a `Requirement`.
     ///
     /// # Examples
     ///
     /// ```
-    /// let req = heroku_go_utils::vrs::Requirement::parse("~1.0").unwrap();
+    /// use heroku_go_utils::vrs::VersionRequirement;
+    /// let req = heroku_go_utils::vrs::GoRequirement::parse("~1.0").unwrap();
     /// ```
     ///
     /// # Errors
     /// Invalid semver requirement `&str` like ">< 1.0", ".1.0", "!=4", etc.
     /// will return an error.
-    pub fn parse(input: &str) -> Result<Self, RequirementParseError> {
-        Ok(semver::VersionReq::parse(input).map(Self)?)
+    fn parse(input: &str) -> Result<Self, RequirementParseError>
+    where
+        Self: std::marker::Sized;
+}
+
+impl VersionRequirement<GoVersion> for GoRequirement {
+    fn satisfies<'a>(&self, version: &GoVersion) -> bool {
+        self.0.matches(&version.0)
     }
 
     /// Parses a go version requirement `&str` as a `Requirement`
     ///
     /// # Examples
     ///
     /// ```
-    /// let req = heroku_go_utils::vrs::Requirement::parse_go("go1.0").unwrap();
+    /// use heroku_go_utils::vrs::VersionRequirement;
+    /// let req = heroku_go_utils::vrs::GoRequirement::parse("go1.0").unwrap();
     /// ```
     ///
     /// # Errors
     /// Invalid semver requirement `&str` like ">< 1.0", ".1.0", "!=4", etc.
     /// will return an error.
-    pub fn parse_go(go_req: &str) -> Result<Self, RequirementParseError> {
-        let stripped_req = go_req
+    fn parse(input: &str) -> Result<Self, RequirementParseError> {
+        input
             .strip_prefix("go")
-            .map_or(Cow::Borrowed(go_req), |req| Cow::Owned(format!("={req}")));
-        Self::parse(&stripped_req)
-    }
-
-    /// Determines if a `&Version` satisfies a `Requirement`
-    #[must_use]
-    pub fn satisfies(&self, version: &Version) -> bool {
-        self.0.matches(&version.0)
-    }
-}
-
-impl TryFrom<String> for Requirement {
-    type Error = RequirementParseError;
-    fn try_from(val: String) -> Result<Self, Self::Error> {
-        Requirement::parse(&val)
+            .map_or(parse_semver_requirement(input), |req| {
```

Comment:
Also, this should probably use `map_or_else` for lazy evaluation of the default value.

Reply:
Changed this in https://github.com/heroku/buildpacks-go/pull/245/commits/11035b41e2c70ead76f7b92e94f6a50ff63583cf